contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,430 @@
+    /// @dev Max tick for full range with tick spacing of 60
+    int24 internal constant MAX_TICK = -MIN_TICK;
+
+    uint256 public constant MINIMUM_LIQUIDITY = 10 ** 3;
+
+    struct CallbackData {
+        address sender;
+        IPoolManager.PoolKey key;
+        IPoolManager.ModifyPositionParams params;
+        bool rebalance;
+    }
+
+    struct PoolInfo {
+        uint128 liquidity;
+        // the fee growth of the aggregate position as of the last action on the individual position
+        uint256 feeGrowthInside0LastX128;
+        uint256 feeGrowthInside1LastX128;
+        // how many uncollected tokens are owed to the position, as of the last computation
+        uint128 tokensOwed0;
+        uint128 tokensOwed1;
+        uint256 blockNumber;
+        address liquidityToken;
+    }
+
+    mapping(PoolId => PoolInfo) public poolToInfo;
+
+    constructor(IPoolManager _poolManager) BaseHook(_poolManager) {}
+
+    modifier ensure(uint256 deadline) {
+        require(deadline >= block.timestamp, "Expired");
+        _;
+    }
+
+    function balanceOf(Currency currency, address user) internal view returns (uint256) {
+        if (currency.isNative()) {
+            return user.balance;
+        } else {
+            return IERC20Minimal(Currency.unwrap(currency)).balanceOf(user);
+        }
+    }
+
+    function modifyPosition(IPoolManager.PoolKey memory key, IPoolManager.ModifyPositionParams memory params)
+        internal
+        returns (BalanceDelta delta)
+    {
+        // msg.sender is the test contract (aka whoever called addLiquidity/removeLiquidity)
+
+        delta = abi.decode(poolManager.lock(abi.encode(CallbackData(msg.sender, key, params, false))), (BalanceDelta));
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+        }
+    }
+
+    function hookModifyPosition(IPoolManager.PoolKey memory key, IPoolManager.ModifyPositionParams memory params)
+        internal
+        returns (BalanceDelta delta)
+    {
+        IERC20Minimal(Currency.unwrap(key.currency0)).approve(address(this), type(uint256).max);
+        IERC20Minimal(Currency.unwrap(key.currency1)).approve(address(this), type(uint256).max);
+
+        delta = abi.decode(poolManager.lock(abi.encode(CallbackData(address(this), key, params, true))), (BalanceDelta));
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+        }
+    }
+
+    function lockAcquired(uint256, bytes calldata rawData) external override returns (bytes memory) {
+        require(msg.sender == address(poolManager));
+
+        CallbackData memory data = abi.decode(rawData, (CallbackData));
+
+        BalanceDelta delta = poolManager.modifyPosition(data.key, data.params);
+
+        // check if we are inputting liquidity for token0
+        if (delta.amount0() > 0) {
```

Comment:
I think instead of having this big if-statement, we have two functions (or maybe just two paths here) one for addingLiquidity, and one for removingLiquidity. We do not need to know from here whether it is a rebasing task, (note: rebasing logic is exactly the same as non-rebasing logic (i.e. lines 126/132).

If `data.params.liquidityDelta` is positive, we know all deltas are the same and they can fit under one it-statement, likewise for negative, and we don't need any conditionals for rebalancing.